libelement/src/object_model/intrinsics/intrinsic_constructor_num.cpp

```diff
@@ -17,13 +17,32 @@ object_const_shared_ptr intrinsic_constructor_num::call(
     const source_information& source_info) const
 {
     auto expr = std::dynamic_pointer_cast<const instruction>(compiled_args[0]);
-
-    if (!expr)
-        return std::make_shared<const element::error>(fmt::format("Argument to Num was '{}' which is invalid. Must be Num or Bool.", compiled_args[0]->to_string()), ELEMENT_ERROR_CONSTRAINT_NOT_SATISFIED, source_info);
-
-    expr->actual_type = type::num.get();
-
-    return evaluate(context, std::move(expr));
+    
+    //If it's not an instruction then we definitely can't convert it, and if it's not a Num/Bool type then it's not valid.
+    if (!expr || !expr->actual_type) {
+        return std::make_shared<const element::error>(
+            fmt::format("Argument to intrinsic 'Num' was '{}' which is invalid. Must be 'Num' or 'Bool'",
+                compiled_args[0]->to_string()), ELEMENT_ERROR_CONSTRAINT_NOT_SATISFIED, source_info);
+    }
+
+    //If it's already a Num, we don't need to do anything
+    if (expr->actual_type == type::num.get())
+        return expr;
+
+    //If it's a Bool, we can wrap it in a IF to convert it to a number
+    if (expr->actual_type == type::boolean.get()) {
```

Comment:
Is the boolean expr's value guaranteed to come out as 0 or 1 here? If so, this feels like a pretty heavy-handed way of turning it into a `num`... but if it could be any value and we need to check it according to the normal boolean rules, then there might be no better way of doing it.

Reply:
yeah 😢 inputs could be a non-bool value at least. sent ya the thought process behind it and some other possible options